### PR TITLE
[api][webui] Add validation for Comment type

### PR DESCRIPTION
**Pull request by Ana06**
The column `type` of `Comment` is reserved for storing the the class in case of inheritance and it is `nil` otherwise. As we validate that it is present it can not be `nil`. So that means that it is not possible to create a `Comment `directly, only instances of `Comment` children can be created. In case that we try to create an` Comment` with a non `nil` type which is not a name of any of the `Comment `children the following error is raised in the database level:

`ActiveRecord::SubclassNotFound: The single-table inheritance mechanism failed to locate the subclass: 'foo'. This error is raised because the column 'type' is reserved for storing the class in case of inheritance. Please rename this column if you didn't intend it to be used for storing the inheritance class or overwrite Parent.inheritance_column to use another column for that information.`

It is not obvious that an instance of `Comment` can not be directly created and that the type is used for the single table inheritance, so the comment and the validation make it much more clear.

It is also a good idea to check it in a Rails validation because for example it won't fail in factories, like happens in https://github.com/openSUSE/open-build-service/pull/2247 where a factory for `comment` is created with a random word as type, which would raise an exception in Rails but works perfectly in a factory.

### CODE REVIEW

File: src/api/app/models/comment.rb [hunk 1]
```diff
@@ -9,6 +9,10 @@ class Comment < ApplicationRecord
```
Comment (bgeuken) [inline src/api/app/models/comment.rb]: regarding the commit message... There is a typo in the first sentence "the the class". And a comma in the second sentence would be nice:-)

**Comment by bgeuken** [comment]
Just a wild idea, but how about making Comment an abstract class then?

File: src/api/app/models/comment.rb [hunk 2]
```diff
@@ -75,4 +79,10 @@ def blank_or_destroy
   def destroy
     super
   end
+
+  private
+
+  def check_is_child
+    errors[:type] << "is reserved for storing the inheritance class which was not found" unless type.safe_constantize.try(:superclass) == Comment
```
Comment (Ana06) [inline src/api/app/models/comment.rb]: According to [Ruby style guide](https://github.com/bbatsov/ruby-style-guide#if-as-a-modifier) it should be in one line:

> Favor modifier if/unless usage when you have a single-line body

**Comment by Ana06** [comment]
I though that making the class `Comment `abstract was not a good idea because then we will store every kind of comment in a different table instead of in the same one like it is doing now. But I have reviewed the code and it seems that storing it this way would be better because we never query comments of different types.

Do I close this and open a PR to make the class abstract or do we want to merge this until I open the other PR?

File: src/api/app/models/comment.rb (hunk 2 above)
Comment (coolo) [inline src/api/app/models/comment.rb]: that's a guide, not a law. And if you cross 80-100 characters with it, you have to ignore that part. Your line is 143 chars and that requires quite some scrolling to understand/review.

There are some creative ways to use the post-if without making the line too long, e.g. 
```
def add_type_error
  errors[:type] << "is reserved for storing the inheritance class which was not found"
end

def check_is_child
  add_type_error unless type.safe_constantize.try(:superclass) == Comment
end
```

But I would just do what @bgeuken suggests.

Comment (Ana06) [inline src/api/app/models/comment.rb]: ok! :+1:

**Comment by bgeuken** [comment]
@Ana06 I think we can merge this one for now. We can make that class abstract later on.

**Comment by marcus-h** [comment]
On 2016-10-24 01:44:09 -0700, Stephan Kulow wrote:
> +
> +  private
> +
> +  def check_is_child
> +    errors[:type] << "is reserved for storing the inheritance class which was not found" unless type.safe_constantize.try(:superclass) == Comment
> 
> that's a guide, not a law. And if you cross 80-100 characters with it, you have to ignore that part. Your line is 143 chars and that requires quite some scrolling to understand/review.
> 
> There are some creative ways to use the post-if without making the line too long, e.g. 
> ```
> def add_type_error
>   errors[:type] << "is reserved for storing the inheritance class which was not found"
> end
> 
> def check_is_child
>   add_type_error unless type.safe_constantize.try(:superclass) == Comment
> end
> ```
> 
Introducing/Spending a method for this is really creative;)
Something less creative:

msg = "is reserved for storing the inheritance class which was not found"
errors[:type] << msg unless type.safe_constantize.try(:superclass) == Comment

**Comment by coolo** [comment]
Method calls are suprisingly cheap compared to assigning unused variables unused strings :)

**Pull request merged**